### PR TITLE
Add validation for decimal values at runtime

### LINKED ISSUE #33805: decimal operations give incorrect results
**Description:**
$subject

**Steps to reproduce:**
```
import ballerina/io;
public function main() {
    io:println(1e-6143 + 1e-6143); // 2e-6143
    io:println(-1e-6143 + 1e-6143); // 0
    io:println(9.999999999999999999999999999999998e6144 + 0.000000000000000000000000000000001e6144); //  9.999999999999999999999999999999999E+6144
    io:println(-9.999999999999999999999999999999998E6144 + -0.000000000000000000000000000000001E6144); //  -9.999999999999999999999999999999999E+6144
    io:println(-9.999999999999999999999999999999999E6144 + 0.000000000000000000000000000000001E6144); //  -9.999999999999999999999999999999998E+6144
    io:println(2E-6143 + 1E-6143); //  3E-6143
    io:println(9E-6143 + 1E-6143); //  1.0E-6142
}
```

The expected output is commented in the code, actual values are
```
0.0
0.0
Infinity
-Infinity
NaN
0.0
0.0
```

**Affected Versions:**
swan-lake beta4 rc1

### PR DESCRIPTION
**Pull request by gabilang**
Fixes #33805

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Comment by KavinduZoysa** [comment]
@gabilang, does this PR cover underflows?

**Comment by KavinduZoysa** [comment]
Shall we add a test https://github.com/ballerina-platform/ballerina-lang/issues/33805#issuecomment-983276303

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/values/DecimalValue.java [hunk 1]
```diff
@@ -90,6 +101,14 @@ public DecimalValue(String value, DecimalValueKind valueKind) {
         this.valueKind = valueKind;
     }
 
+    private static BigDecimal getValidDecimalValue(BigDecimal bd) {
```
Comment (HindujaB) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/values/DecimalValue.java:104]: Shall we move the hex value check inside this method as well?
So that we can keep all the validations in one method.

File: langlib/langlib-test/src/test/resources/test-src/decimallib_test.bal [hunk 2]
```diff
@@ -396,6 +396,24 @@ function testFromStringFunctionWithInvalidValues() {
         assertEquality("{ballerina/lang.decimal}NumberParsingError", a1.message());
         assertEquality("'string' value '' cannot be converted to 'decimal'", <string> checkpanic a1.detail()["message"]);
     }
+
+    a1 = decimal:fromString("1e-6143");
+    assertEquality(false, a1 is error);
+    assertEquality(1e-6143d, checkpanic a1);
+
+    a1 = trap decimal:fromString("9.999999999999999999999999999999999E6145");
```
Comment (HindujaB) [inline langlib/langlib-test/src/test/resources/test-src/decimallib_test.bal:404]: What happens with the hardcoded decimal values that are out of range?
```ballerina
    decimal x =  9.999999999999999999999999999999999E6145;
```
currently, it didn't throw any exceptions.

File: tests/jballerina-unit-test/src/test/resources/test-src/expressions/typecast/type-casting.bal [hunk 3]
```diff
@@ -28,6 +28,45 @@ function stringtoint(string value) returns int|error {
...
+    int|error result = trap <int> d3;
+    assertEquality(true, result is error);
+    error err = <error> result;
+    assertEquality("{ballerina}NumberConversionError", err.message());
+    assertEquality("'decimal' value '-9223372036854775808.9' cannot be converted to 'int'",
+                    checkpanic <string|error> err.detail()["message"]);
+}
+
+function testDecimalToFloatCasting() {
+    decimal d = 0.0000000000000009e+308;
```
Comment (HindujaB) [inline tests/jballerina-unit-test/src/test/resources/test-src/expressions/typecast/type-casting.bal:51]: Shall we add tests for negative values as well?

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/values/DecimalValue.java [hunk 4]
```diff
@@ -50,26 +51,34 @@ public class DecimalValue implements SimpleValue, BDecimal {
     private static final String NEG_INF_STRING = "-" + INF_STRING;
     private static final String NAN = "NaN";
 
+    private static final BigDecimal DECIMAL_MAX =
+            new BigDecimal("9.999999999999999999999999999999999e6144", MathContext.DECIMAL128);
+
+    private static final BigDecimal DECIMAL_MIN =
+            new BigDecimal("-9.999999999999999999999999999999999e6144", MathContext.DECIMAL128);
+
```
Comment (HindujaB) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/values/DecimalValue.java:59]: Shall we remove the additional line here?

File: langlib/langlib-test/src/test/resources/test-src/decimallib_test.bal (hunk 2 above)
Comment (gabilang) [reply langlib/langlib-test/src/test/resources/test-src/decimallib_test.bal:404]: It should result in compilation error and being addressed in https://github.com/ballerina-platform/ballerina-lang/pull/37823

**Review by KavinduZoysa — APPROVED** [review]

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/values/DecimalValue.java (hunk 1 above)
Comment (gabilang) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/values/DecimalValue.java:101]: This method has been used in some other places where we need to pass `BigDecimal` value as an argument. We need to pass a hexadecimal string value to the method `hexToDecimalFloatingPointNumber`. So moving it here may need another method which needs to get BigDecimal as an argument. IMO, keeping as it is may be more convenient. WDYT?

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/values/DecimalValue.java [hunk 5]
```diff
@@ -57,19 +62,20 @@ public class DecimalValue implements SimpleValue, BDecimal {
...
             this.valueKind = DecimalValueKind.ZERO;
         }
     }
 
     public DecimalValue(String value) {
         // Check whether the number provided is a hexadecimal value.
+        BigDecimal bd;
         if (isHexValueString(value)) {
-            this.value = hexToDecimalFloatingPointNumber(value);
+            bd = hexToDecimalFloatingPointNumber(value);
```
Comment (warunalakshitha) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/values/DecimalValue.java:75]: Any idea why this line not covered?

Comment (gabilang) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/values/DecimalValue.java:75]: It looks like previously no tests with hex string have been added to cover the lines of the method `hexToDecimalFloatingPointNumber`. Will try to add some cases for that.

Comment (gabilang) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/values/DecimalValue.java:75]: According to [spec](https://ballerina.io/spec/lang/2022R2/#lang.decimal) `decimal:String()` is not going to accept hexadecimal strings as arguments and it has been addressed with the issue https://github.com/ballerina-platform/ballerina-lang/issues/33406. It looks like JsonParser also need not to be read and convert hexadecimal strings to decimal values. So the method added to convert hexadecimal string to BigDecimal value in `DecimalValue.java` can be safely removed.

**Review by warunalakshitha — APPROVED** [review]
LG
